File: javascript/reading-file-exercise/tests/test_ReadingFile.js
/**
 * test_ReadingFile.js — Unit tests for parseCsv and loadCustomers.
 * Runs in the browser via tests/test-runner.html.
 */

(function () {
    const results = [];
    let passCount = 0;
    let failCount = 0;

    function assert(condition, message) {
        if (!condition) throw new Error(message || "Assertion failed");
    }

    function assertEqual(actual, expected, message) {
        if (actual !== expected) {
            throw new Error(
                (message || "assertEqual") +
                    ` — expected: ${JSON.stringify(expected)}, received: ${JSON.stringify(actual)}`
            );
        }
    }

    function assertThrows(fn, expectedMessage) {
        let threw = false;
        try {
            fn();
        } catch (e) {
            threw = true;
            if (expectedMessage && !e.message.includes(expectedMessage)) {
                throw new Error(
                    `Expected error with "${expectedMessage}", got: "${e.message}"`
                );
            }
        }
        if (!threw) {
            throw new Error(`Expected to throw error${expectedMessage ? ': "' + expectedMessage + '"' : ""}`);
        }
    }

    async function assertRejects(fn, expectedMessage) {
        let threw = false;
        try {
            await fn();
        } catch (e) {
            threw = true;
            if (expectedMessage && !e.message.includes(expectedMessage)) {
                throw new Error(
                    `Expected error with "${expectedMessage}", got: "${e.message}"`
                );
            }
        }
        if (!threw) {
            throw new Error(`Expected to reject${expectedMessage ? ': "' + expectedMessage + '"' : ""}`);
        }
    }

    async function runTest(name, fn) {
        try {
            await fn();
            results.push({ name, status: "PASSED" });
            passCount++;
        } catch (e) {
            results.push({ name, status: "FAILED", error: e.message });
            failCount++;
        }
    }

    function renderResults() {
        const container = document.getElementById("results");
        const summary = document.createElement("h2");
        summary.textContent = `${passCount} passed, ${failCount} failed — ${results.length} total`;
        summary.style.color = failCount > 0 ? "#b71c1c" : "#2e7d32";
        container.appendChild(summary);

        results.forEach((r) => {
            const div = document.createElement("div");
            div.style.padding = "6px 10px";
            div.style.margin = "4px 0";
            div.style.borderRadius = "4px";
            div.style.fontFamily = "monospace";

            if (r.status === "PASSED") {
                div.style.background = "#e8f5e9";
                div.style.color = "#2e7d32";
                div.textContent = `✓ ${r.name}`;
            } else {
                div.style.background = "#ffebee";
                div.style.color = "#b71c1c";
                div.textContent = `✗ ${r.name} — ${r.error}`;
            }
            container.appendChild(div);
        });
    }

    // ================================================================
    // TESTS — parseCsv
    // ================================================================

    async function parseCsv_returns_columns_and_rows_from_valid_csv() {
        const csv = "Name,Age,City\nAlice,30,Sao Paulo\nBob,25,Rio";
        const data = parseCsv(csv);

        assertEqual(data.columns.length, 3);
        assertEqual(data.columns[0], "Name");
        assertEqual(data.columns[1], "Age");
        assertEqual(data.columns[2], "City");
        assertEqual(data.rows.length, 2);
        assertEqual(data.rows[0]["Name"], "Alice");
        assertEqual(data.rows[1]["City"], "Rio");
    }

    async function parseCsv_throws_error_for_empty_csv() {
        assertThrows(() => parseCsv(""), "Empty CSV or missing header");
    }

    async function parseCsv_throws_error_for_whitespace_only_csv() {
        assertThrows(() => parseCsv("   \n  "), "Empty CSV or missing header");
    }

    async function parseCsv_returns_zero_rows_for_header_only_csv() {
        const csv = "Name,Age,City";
        const data = parseCsv(csv);

        assertEqual(data.columns.length, 3);
        assertEqual(data.rows.length, 0);
    }

    async function parseCsv_skips_rows_with_mismatched_column_count() {
        const csv = "A,B\n1,2\n3\n4,5";
        const data = parseCsv(csv);

        assertEqual(data.rows.length, 2);
        assertEqual(data.rows[0]["A"], "1");
        assertEqual(data.rows[1]["A"], "4");
    }

    async function parseCsv_trims_whitespace_around_values() {
        const csv = " Name , Age \n  Alice , 30 ";
        const data = parseCsv(csv);

        assertEqual(data.columns[0], "Name");
        assertEqual(data.columns[1], "Age");
        assertEqual(data.rows[0]["Name"], "Alice");
        assertEqual(data.rows[0]["Age"], "30");
    }

    async function parseCsv_handles_crlf_line_breaks() {
        const csv = "X,Y\r\n1,2\r\n3,4";
        const data = parseCsv(csv);

        assertEqual(data.rows.length, 2);
        assertEqual(data.rows[0]["X"], "1");
    }

    async function parseCsv_treats_empty_value_as_empty_string() {
        const csv = "A,B\n,hello";
        const data = parseCsv(csv);

        assertEqual(data.rows[0]["A"], "");
        assertEqual(data.rows[0]["B"], "hello");
    }

    async function parseCsv_parses_quoted_fields_with_commas() {
        const csv = 'Name,Address\n"Alice","Avenue, 123"';
        const data = parseCsv(csv);

        assertEqual(data.rows.length, 1);
        assertEqual(data.rows[0]["Name"], "Alice");
        assertEqual(data.rows[0]["Address"], "Avenue, 123");
    }

    async function parseCsv_parses_escaped_quotes_inside_quoted_field() {
        const csv = 'Name,Note\n"Bob","He said ""hello"""';
        const data = parseCsv(csv);

        assertEqual(data.rows.length, 1);
        assertEqual(data.rows[0]["Note"], 'He said "hello"');
    }

    async function parseCsv_throws_error_for_unclosed_quoted_field() {
        assertThrows(() => parseCsv('A,B\n"1,2'), "missing closing quote");
    }

    async function parseCsv_parses_multiline_quoted_field() {
        const csv = 'Name,Bio\n"Alice","Line 1\nLine 2\nLine 3"\n"Bob","Simple"';
        const data = parseCsv(csv);

        assertEqual(data.rows.length, 2);
        assertEqual(data.rows[0]["Name"], "Alice");
        assertEqual(data.rows[0]["Bio"], "Line 1\nLine 2\nLine 3");
        assertEqual(data.rows[1]["Name"], "Bob");
        assertEqual(data.rows[1]["Bio"], "Simple");
    }

    async function parseCsv_parses_multiline_quoted_field_with_crlf() {
        const csv = 'A,B\r\n"hello\r\nworld",2\r\nx,y';
        const data = parseCsv(csv);

        assertEqual(data.rows.length, 2);
        assertEqual(data.rows[0]["A"], "hello\r\nworld");
        assertEqual(data.rows[0]["B"], "2");
        assertEqual(data.rows[1]["A"], "x");
    }

    async function parseCsv_parses_multiline_with_commas_and_quotes() {
        const csv = 'Name,Address\n"Alice","Avenue, 123\nApt ""B"""';
        const data = parseCsv(csv);

        assertEqual(data.rows.length, 1);
        assertEqual(data.rows[0]["Address"], 'Avenue, 123\nApt "B"');
    }

    // ================================================================
    // TESTS — loadCustomers
    // ================================================================

    async function loadCustomers_loads_real_csv_successfully() {
        const data = await loadCustomers();

        assert(data.columns.length > 0, "Should have at least one column");
        assert(data.rows.length > 0, "Should have at least one row");
        assert(typeof data.rows[0] === "object", "Each row should be an object");
    }

    async function loadCustomers_returns_columns_as_row_keys() {
        const data = await loadCustomers();

        data.columns.forEach((col) => {
            assert(col in data.rows[0], `Column "${col}" should exist in the first row`);
        });
    }

    async function loadCustomers_throws_error_for_missing_url() {
        await assertRejects(
            () => loadCustomers("./nonexistent-file.csv"),
            "Error loading file"
        );
    }

    // ================================================================
    // RUN
    // ================================================================

    async function runAll() {
        await runTest("parseCsv returns columns and rows from valid CSV", parseCsv_returns_columns_and_rows_from_valid_csv);
        await runTest("parseCsv throws error for empty CSV", parseCsv_throws_error_for_empty_csv);
        await runTest("parseCsv throws error for whitespace-only CSV", parseCsv_throws_error_for_whitespace_only_csv);
        await runTest("parseCsv returns zero rows for header-only CSV", parseCsv_returns_zero_rows_for_header_only_csv);
        await runTest("parseCsv skips rows with mismatched column count", parseCsv_skips_rows_with_mismatched_column_count);
        await runTest("parseCsv trims whitespace around values", parseCsv_trims_whitespace_around_values);
        await runTest("parseCsv handles CRLF line breaks", parseCsv_handles_crlf_line_breaks);
        await runTest("parseCsv treats empty value as empty string", parseCsv_treats_empty_value_as_empty_string);
        await runTest("parseCsv parses quoted fields with commas", parseCsv_parses_quoted_fields_with_commas);
        await runTest("parseCsv parses escaped quotes inside quoted field", parseCsv_parses_escaped_quotes_inside_quoted_field);
        await runTest("parseCsv throws error for unclosed quoted field", parseCsv_throws_error_for_unclosed_quoted_field);
        await runTest("parseCsv parses multiline quoted field", parseCsv_parses_multiline_quoted_field);
        await runTest("parseCsv parses multiline quoted field with CRLF", parseCsv_parses_multiline_quoted_field_with_crlf);
        await runTest("parseCsv parses multiline with commas and quotes", parseCsv_parses_multiline_with_commas_and_quotes);
        await runTest("loadCustomers loads real CSV successfully", loadCustomers_loads_real_csv_successfully);
        await runTest("loadCustomers returns columns as row keys", loadCustomers_returns_columns_as_row_keys);
        await runTest("loadCustomers throws error for missing URL", loadCustomers_throws_error_for_missing_url);

        renderResults();
    }

    if (document.readyState === "loading") {
        document.addEventListener("DOMContentLoaded", runAll);
    } else {
        runAll();
    }
})();

File: python/reading-file-exercise/tests/test_reading_file.py
import csv
from pathlib import Path

import pytest

from ReadingFile import get_csv_path, load_customers, parse_csv_text


def test_get_csv_path_points_to_existing_customers_file():
    csv_path = get_csv_path()

    assert csv_path.name == "customers-10000.csv"
    assert csv_path.exists()


def test_load_customers_reads_sample_csv_successfully():
    customers = load_customers(get_csv_path())

    assert isinstance(customers, list)
    assert len(customers) > 0
    assert isinstance(customers[0], dict)
    assert len(customers[0]) > 0


def test_load_customers_raises_file_not_found_for_missing_path(tmp_path: Path):
    missing_path = tmp_path / "missing.csv"

    with pytest.raises(FileNotFoundError):
        load_customers(missing_path)


def test_load_customers_raises_value_error_for_empty_csv(tmp_path: Path):
    empty_path = tmp_path / "empty.csv"
    empty_path.write_text("", encoding="utf-8")

    with pytest.raises(ValueError, match="Empty CSV or missing header"):
        load_customers(empty_path)


def test_load_customers_raises_unicode_decode_error_for_invalid_encoding(tmp_path: Path):
    invalid_encoding_path = tmp_path / "invalid-encoding.csv"
    invalid_encoding_path.write_bytes(b"\x80\x81\x82")

    with pytest.raises(UnicodeDecodeError):
        load_customers(invalid_encoding_path)


def test_load_customers_raises_permission_error_when_open_fails(tmp_path: Path, monkeypatch):
    protected_path = tmp_path / "protected.csv"
    protected_path.write_text("name\nAlice\n", encoding="utf-8")

    def raise_permission_error(*args, **kwargs):
        raise PermissionError("access denied")

    monkeypatch.setattr(Path, "open", raise_permission_error)

    with pytest.raises(PermissionError, match="Permission denied reading file"):
        load_customers(protected_path)


def test_load_customers_raises_value_error_when_csv_reader_fails(tmp_path: Path, monkeypatch):
    valid_path = tmp_path / "valid.csv"
    valid_path.write_text("name\nAlice\n", encoding="utf-8")

    def raise_csv_error(*args, **kwargs):
        raise csv.Error("bad csv")

    monkeypatch.setattr(csv, "DictReader", raise_csv_error)

    with pytest.raises(ValueError, match="Failed to process CSV"):
        load_customers(valid_path)


# ================================================================
# TESTS — parse_csv_text
# ================================================================


def test_parse_csv_text_returns_columns_and_rows_from_valid_csv():
    data = parse_csv_text("Name,Age,City\nAlice,30,Sao Paulo\nBob,25,Rio")

    assert data["columns"] == ["Name", "Age", "City"]
    assert len(data["rows"]) == 2
    assert data["rows"][0]["Name"] == "Alice"
    assert data["rows"][1]["City"] == "Rio"


def test_parse_csv_text_raises_value_error_for_empty_string():
    with pytest.raises(ValueError, match="Empty CSV or missing header"):
        parse_csv_text("")


def test_parse_csv_text_raises_value_error_for_whitespace_only():
    with pytest.raises(ValueError, match="Empty CSV or missing header"):
        parse_csv_text("   \n  ")


def test_parse_csv_text_returns_zero_rows_for_header_only_csv():
    data = parse_csv_text("Name,Age,City")

    assert data["columns"] == ["Name", "Age", "City"]
    assert len(data["rows"]) == 0


def test_parse_csv_text_parses_quoted_fields_with_commas():
    data = parse_csv_text('Name,Address\n"Alice","Avenue, 123"')

    assert len(data["rows"]) == 1
    assert data["rows"][0]["Address"] == "Avenue, 123"


def test_parse_csv_text_parses_escaped_quotes_inside_quoted_field():
    data = parse_csv_text('Name,Note\n"Bob","He said ""hello"""')

    assert data["rows"][0]["Note"] == 'He said "hello"'


def test_parse_csv_text_parses_multiline_quoted_field():
    data = parse_csv_text('Name,Bio\n"Alice","Line 1\nLine 2\nLine 3"\n"Bob","Simple"')

    assert len(data["rows"]) == 2
    assert data["rows"][0]["Bio"] == "Line 1\nLine 2\nLine 3"
    assert data["rows"][1]["Name"] == "Bob"


def test_parse_csv_text_parses_multiline_quoted_field_with_crlf():
    data = parse_csv_text('A,B\r\n"hello\r\nworld",2\r\nx,y')

    assert len(data["rows"]) == 2
    assert data["rows"][0]["A"] == "hello\r\nworld"
    assert data["rows"][0]["B"] == "2"


def test_parse_csv_text_parses_multiline_with_commas_and_quotes():
    data = parse_csv_text('Name,Address\n"Alice","Avenue, 123\nApt ""B"""')

    assert len(data["rows"]) == 1
    assert data["rows"][0]["Address"] == 'Avenue, 123\nApt "B"'

